react/features/polls/components/PollDialog.js

```diff
@@ -0,0 +1,363 @@
+    isPollRunning: boolean,
+
+    /**
+     * Invoked to obtain translated strings.
+     */
+    t: Function,
+
+    /**
+     * Current user id.
+     */
+    userID: string
+};
+
+type State = {
+
+    /**
+     * Available poll choices.
+     */
+    choices: Object,
+
+    /**
+     * New Poll Object
+     */
+    poll: Object,
+
+    /**
+     * Poll question.
+     */
+    question: Object
+};
+
+/**
+ * Polls main dialog view component.
+ */
+class PollDialog extends Component<Props, State> {
+
+    /**
+     * Constructor.
+     *
+     * @inheritdoc
+     */
+    constructor(props: Props) {
+        super(props);
+
+        this._closeDialog = this._closeDialog.bind(this);
+        this._onSubmit = this._onSubmit.bind(this);
+        this._addChoice = this._addChoice.bind(this);
+        this._removeChoice = this._removeChoice.bind(this);
+        this._onQuestionTextChange = this._onQuestionTextChange.bind(this);
+        this._createNewPoll = this._createNewPoll.bind(this);
+        this._choiceTextChange = this._choiceTextChange.bind(this);
+        this._vote = this._vote.bind(this);
+        this._endPoll = this._endPoll.bind(this);
+        this._clearState = this._clearState.bind(this);
+
+        const initialState = this._clearState();
+
+        this.state = {
+            ...initialState
+        };
+    }
+
+    _closeDialog: () => void;
+
+    /**
+     * Callback invoked to close the dialog without saving changes.
+     *
+     * @private
+     * @returns {void}
+     */
+    _closeDialog() {
+        this.props.dispatch(hideDialog());
+    }
+
+    /**
+     * Component render method.
+     *
+     * @inheritdoc
+     */
+    render() {
+        const { isPollRunning, userID } = this.props;
+        const title = isPollRunning ? 'dialog.endPoll' : 'dialog.startPoll';
+        const tabs = isPollRunning ? [ {
+            component: VoteForm,
+            label: 'polls.vote',
+            props: {
+                userID,
+                onVoteClicked: this._vote
+            },
+            submit: null
+        } ] : [ {
+            component: PollCreateForm,
+            label: 'polls.create',
+            props: {
+                choices: this.state.choices,
+                onAddChoice: this._addChoice,
+                onChoiceTextChange: this._choiceTextChange,
+                onRemoveChoice: this._removeChoice,
+                onQuestionTextChange: this._onQuestionTextChange
+            },
+            propsUpdateFunction: (oldProps, newProps) => {
+                return { ...newProps };
+            },
+            submit: null
+        } ];
+
+        tabs.push({
+            component: PollResultsForm,
+            label: 'polls.results',
+            props: {},
+            submit: null
+        });
+
+        return (
+            <DialogWithTabs
+                closeDialog = { this._closeDialog }
+                cssClassName = { 'polls-dialog' }
+                key = { isPollRunning.toString() }
```

Comment:
Do you need this key?

Reply:
Yes, the problem is that DialogWithTabs is designed in a way that it doesn't expect dynamic tabs. It saves the props of the tabs in it's constructor. My use case is that I use dynamic tabs to toggle the Create and Vote forms in place.

Since Dialog appears in same place, react doesn't re-construct it and the tab props for the Vote form is not propagated accordingly. This key forces react to recreate the Dialog.